### ISSUE TITLE
Getting "control.registerOnChange is not a function" after changing form

### ISSUE THREAD

**Issue by natcohen** [angular/angular#13788]
**I'm submitting a ...**  (check one with "x")
```
[ x ] bug report => search github for a similar issue or PR before submitting
[   ] feature request
[   ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

After patching a form of type `FormArray` (`form.patchValue(myArray);`) with `FormGroup` items,  `[formGroup]='form'` in my html throws `control.registerOnChange is not a function`. I patch it when I initialize my app and everything works fine but when I change the form with another `FormArray` (I tried with the same array), I'm getting this error. After further investigation (and hours trying to understand what was going on), it seems that the error comes from line 1634 of `forms.umd.js` when it does `control.registerOnChange(function (newValue, emitModelEvent) {`.

When it works, `control` (of `control.registerOnChange`) is of form `FormControl` while when it fails, it is of form `FormGroup`. The problem is that my `FormArray` is somewhat deep and the first time I patch it, it goes at the leaf of the branches to reach `FormControl` while the second time it goes to the first branch which in my case is a `FormGroup`.

The problem is that I don't know why it doesn't go to the leaf the second time. I use the same functions and methodology to patch the values and the `FormArray` looks exactly the same in both cases!

**Expected behavior**

It should not throw any error...

**Minimal reproduction of the problem with instructions**

I couldn't make a minimal demo but would definitely share the functions I use to create the form if needed.

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

NodeJS 7.4.0 running on Windows 10

* **Angular version:** 2.4.7

* **Browser:** Chrome 55 64bit

* **Language:** Typescript

**Comment by DzmitryShylovich** [comment]
pls add a small demo in plunkr otherwise it's not actionable

**Comment by natcohen** [comment]
Ok, here is the bug...

https://plnkr.co/edit/MolZlFprywZ1nSfRhujb?p=preview

**Comment by natcohen** [comment]
@DzmitryShylovich and @matsko Can someone confirm it's a bug? I'm stuck and would like to move on...

**Comment by matsko** [comment]
cc @kara

**Comment by dejour** [comment]
`<input name="answer" type="text" placeholder="Reponse" formControlName="answer"/>` remove formControlName doesn't produce error

**Comment by dejour** [comment]
because answer is a formGroup. key is number based index.

**Comment by natcohen** [comment]
@dejour please look more carefully at the code! There is a *ngIf that displays the group as checkboxes... the problem occurs only when the DOM is updated.

**Comment by natcohen** [comment]
@kara could you please confirm it's a bug? If it's a bad implementation, I need to change that ASAP.

Thanks

**Comment by natcohen** [comment]
I have simplified the plunker: https://plnkr.co/edit/dJ1KyDLFDOOa1FUfWu5d?p=preview

**Comment by dejour** [comment]
after taking a close look, it seems like a bug

**Comment by dejour** [comment]
I added a new form group in the second place in changed template, which is the same index with inital control's index.https://plnkr.co/edit/lqxzPaYAc2iK4uKrvkRO?p=preview

**Comment by dejour** [comment]
@DzmitryShylovich @kara

**Comment by Bretto** [comment]
I have the same problem, I need to add dynamically checkboxes to a form: 
I have tried with formArray but when I add a groupForm to the formArray it breaks. I think there is a bug.  Could someone show how it's supposed to be done: 
https://plnkr.co/edit/Ark5W6bog6BX4yV9ONEd?p=preview

**Comment by dejour** [comment]
https://plnkr.co/edit/PJHbJBYGD74cfPIruAzQ?p=preview @Bretto

**Comment by dejour** [comment]
https://plnkr.co/edit/PJHbJBYGD74cfPIruAzQ?p=preview @Bretto

**Comment by Bretto** [comment]
Thank you very much I wasn't aware of the: ```addControl``` 👍

**Comment by natcohen** [comment]
@matsko @kara and @DzmitryShylovich I know there are a lot of new issues everyday but I've spent two days trying to find this bug... I would expect at least the bug tag or a little more attention! Sorry but I'm a bit frustrated

**Comment by natcohen** [comment]
@vicb and @DzmitryShylovich I'm a little bit desperate... Can someone add (at least) the bug tag?

**Comment by natcohen** [comment]
@DzmitryShylovich and @kara please help with this bug... 3 months and still nothing for a blocking bug!

**Comment by lazaridis-com** [comment]
@natcohen , naturally, you have to find a workaround to your issue. What about those two suggestions:

https://github.com/angular/angular/issues/13788#issuecomment-272398272
https://github.com/angular/angular/issues/13788#issuecomment-274014852

**Comment by ThalesMatoso** [comment]
solved? i have this problem ...

**Comment by jeffgrady** [comment]
#MeToo

**Comment by manisabri** [comment]
I'm having the same problem with dynamically adding custom controls in a formArray on the fly. The workarounds are horrible. 
I can dismantle the custom control and just pass in the formGroup instance and go from there but it makes the code much harder to maintain and doesn't count as encapsulation.

**Comment by manisabri** [comment]
Creating a new formArray and replacing the old one was better workaround for me . So I replaced the below line 
`this.documents.push(this.fb.group(new Document)));`
with: 
```
this.job.documents.push(new Document());
let docs = this.fb.array(this.job['documents'].map(document => new Document(false, null, document)));
this.jobForm.controls.documents = docs;
```

**Comment by natcohen** [comment]
This fix works: https://github.com/ghetolay/angular/commit/535147d5a1261a281b54d9be4a7de65562d8d664

But nobody wants to merge it :(

**Comment by abhijithmannath** [comment]
I am facing issues with dynamically adding to a formarray. Can someone please tell me a workaround ?

**Comment by manisabri** [comment]
I'm using reactive forms and did it like this , hope it makes sense

get discounts(){
    return this.formGroup.get('discounts') as FormArray;
}

addDiscount(){
    let discounts = Object.create(this.discounts.value?this.discounts.value:[]);
    discounts.push(new Discount(false, null, null));
    let discountArr = this.fb.array(discounts.map(discount=> new
Discount(false, null, discount)));
    this.formGroup.setControl('discounts', discountArr);
    this.formGroup.markAsDirty();
}

**Comment by jeffgrady** [comment]
To paraphrase your example, I think I ended up fixing this error this way:

```
let myCtrl: Array<FormControl> = [];
for (let foo of barBaz) {
  myCtrl.push(new FormControl(foo));
}
let discountArr = new FormArray(myCtrl);
...
```

**Comment by natcohen** [comment]
This should have been fixed a long time ago :(

I understand angular's team has a lot to do but this is almost a one year old bug and I'm still having lots of troubles with it...

**Comment by ranglang** [comment]
# ****

**Comment by natcohen** [comment]
@kara @matsko @vicb and @DzmitryShylovich any chance one of you can have a look before the end of the year (Xmas gift :)? It shouldn't take that long to fix, a [solution has already been proposed](https://github.com/ghetolay/angular/commit/535147d5a1261a281b54d9be4a7de65562d8d664)...

**Comment by jotatoledo** [comment]
Any news on this?

**Comment by raysuelzer** [comment]
I take it this is still broken?  Do something like this will not work...

```
...
   <div class="col__" [formGroupName]="i" *ngFor="let tagAndCatControl of groupControl.controls.categoriesAndTags.controls; let i = index;">
                    <my-input [formControl]="tagAndCatControl"></my-input>
```

**Comment by natcohen** [comment]
Still buggy for me...

**Comment by manisabri** [comment]
Maybe it's not broken. I thought it was broken a while back but now the
places that I expected it to work as we do around here got so complicated
that I needed a good  bunch of glue logic to pull it off. So maybe not
broken after all

**Comment by natcohen** [comment]
Almost 3 years and not a word from the Angular team!

It is a shame that the Angular community spends time to find bugs and work on solutions but the team doesn't pay attention to that! What should I do to get it fixed in the next release?

I am currently working on a new project and have moved to React... I love Angular but if we can't get the support we need after 3 years, it makes it unusable!

**Comment by rickyhai11** [comment]
it still happens to me when I try to use nested formArray within a formgroup for the multi-checkboxes use case. This seems to be a basic case but it did not work at all. I am completely disappointed.

**Comment by martinsotirov** [comment]
In my Angular app my form works correctly but I get these `"control.registerOnChange is not a function"` errors every time I update the form. I use ngxsForm so I assume that's the reason I can use the form at all, even though Angular doesn't like the dynamic updating.

My structure is:

```
form = new FormGroup({
  features: new FormArray([]),
}
```

and then the `features` FormArray gets dynamically filled with FormGroups each containing a `value` FormControl where I save my actual values.

**Comment by AndrewKushnir** [comment]
Hi, just want to give an update: we were able to reproduce the issue and I've created a draft PR #40829 with the proposed fix (similar to the fix proposed by @natcohen in PR #27246, thanks @natcohen!). The fix will need to go through the regular code review and testing procedures, but I'd like to ask if anyone who is experiencing this issue could try the fix on their apps and provide some feedback too. 

The "@angular/forms" package (with the fix included) is available at:

https://918438-24195339-gh.circle-artifacts.com/0/angular/forms-pr40829-07bab94bf1.tgz

so you could test it by updating your `package.json` file like this:

```
  ...
  "@angular/forms": "https://918438-24195339-gh.circle-artifacts.com/0/angular/forms-pr40829-07bab94bf1.tgz",
  ...
```

and running `npm install` (more info is available in [DEVELOPERS.md](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#archives-for-each-package)).

Important note: the package is not suitable for production use and will be available for a short period of time.

Thank you.

**Comment by natcohen** [comment]
@AndrewKushnir thank you for finally adding the fix to the official release...but I must say that you guys lost me! Angular is awesome and the concept is fantastic.

Unfortunately if it takes 4 years to merge a fix (to an identified bug) proposed by a user, the project is not worth to be used by production products. I had to move away from this amazing tech because I couldn’t afford to struggle with such issues. You lost a user on this and you probably lost a dozen of potential ones due to the bad press I gave to my dev friends.

Sorry if this post sounds inappropriate but... gosh...4 years?!?!?
